File: Java/queue.java

// ----------------------------------Implementation of Queue Data Structure using Array------------------------------------------

public class queue {
    
  static class Queue{         //----queue class----
        
    static int arr[];         // when we implement a queue using an array then its size is fixed as the size of the array is fixed
    int n;                    // n is size of array 
    static int rear=-1;       // rear is the end of queue, when queue is empty rear is -1
    
        Queue(int size){       // the constructor of queue class which take size from the user as a parameter
            arr=new int[size];
            this.n = size;
        }
    
   
        //---------The method which tells whether the queue is empty or not----------
        public boolean isEmpty() {
            return rear == -1;
        }
    
    
        //---------The method used to add element in the queue---------------
        public void add(int data) {
            if(rear==n-1){
                System.out.println("full queue");
                return;
            }
            rear++;
            arr[rear]= data;
        }
    
    
        //---------The method used to remove the front element from the queue---------------
        public int remove() {
            if (isEmpty()) {
                System.out.println("empty queue");
                return -1;
            }
            int front = arr[0];
            for (int i = 0; i < rear; i++) {
                arr[i]= arr[i+1];
            }
            rear--;
            return front;
        }

    
    
        //---------The method used to get front element of the queue---------------
        public int peek() {
            if (isEmpty()) {
                System.out.println("empty queue");
                return -1;
            }
            return arr[0];
        }

    }
  

  public static void main(String[] args) {   // main function
    Queue q = new Queue(7);                  // object of queue is created having size 7
    q.add(1);                                // add element in queue
    q.add(2);
    q.add(3);
    q.add(4);
    q.add(5);
    q.add(6);
    q.add(7);
    
    while (!q.isEmpty()) {                   // checking queue is empty or not
        System.out.println(q.peek());        // when queue is not empty print its front element
        q.remove();                          // then remove its front element 
    }
   } 
}

File: C++/queue.cpp
#include <iostream>
#include <cstdlib>
#include <string>

using namespace std;

// structure defined
// all functions declared inside the structure
struct queue{
    int size;
    int front;
    int rear;
    int *arr;

    queue(int s){
        size = s;
        arr = new int[size];
        front = -1;
        rear = -1;
    }

    // checks for empty queue
    bool isEmpty(){
        return (front == -1 && rear == -1);
    }

    // checks for a full queue
    bool isFull(){
        return (rear == (size-1) );
    }


    // pushes the elements in queue with parameter n
    void enQueue(int n){
        if(isFull()){
            cout << "The queue is FULL " << endl;
        }
        else{
            if(front == -1){
                front = 0;
            }

            rear++;
            arr[rear] = n;
            
        }
    }

    // tries to remove the element of the queue to which rear points , so does the queue operates (LIFO)
    void deQueue(){
        if(front>rear){
            cout << "The queue is EMPTY " << endl;
        }

        else{
            cout << "The dequeued element is " << arr[rear] << endl;
            front++;
        }
    }

    // prints the elements in the queue 
    void print(){
        if(isEmpty()){
            cout << "The queue is FULL " << endl;
        }

        else{
            cout << "The elements in the queue are: " << endl;
            for(int i = 0 ; i < size ; i++){
                cout << arr[i] << endl;

            }
        }
    }


};



// all functions passed in main
int main(){
    int size;
    cout << "Enter the size of the queue: ";
    cin >> size;

    // structure object defined as q
    queue q(size);

    q.enQueue(21);
    q.enQueue(30);
    q.enQueue(33);
    q.enQueue(81);

    

    q.print();

    q.deQueue();


   return 0;

}



// TIME complexity - O(1) (big-Oh(1)) all the mentioned operations run in constant time.
// SPACE complexity - O(1) (big-Oh(1))